### PR TITLE
OpenCL support for cracking Telegram Desktop passcodes

### PR DESCRIPTION
**Pull request by kholia**
NOTE: This needs to be converted to a pure OpenCL kernel before it can be merged. I might not be able to get this done soon.

@magnumripper I am still weak at writing (and debugging!) low level crypto primitives. Would it be possible for you to port the AES IGE stuff to OpenCL?

### CODE REVIEW

**Comment by kholia** [comment]
CC @MihaZupan for visibility and testing.

**Comment by magnumripper** [comment]
I will have no time in weeks. This isn't rocket science though. These functions in OpenCL are nearly identical to the CPU ones. Here's CPU version of AES_cfb_decrypt (in encfs_common_plug.c):
```
static void AES_cfb_decrypt(AES_KEY *akey, int len, unsigned char *iv,
                            const unsigned char *input, unsigned char *output)
{
        int n = 0;

        while (len--) {
                unsigned char c;

                if (!n)
                        AES_ecb_encrypt(iv, iv, akey, AES_ENCRYPT);

                c = *input++;
                *output++ = c ^ iv[n];
                iv[n] = c;

                n = (n + 1) & 0x0f;
        }
}
```

...and this is our current OpenCL version of the same:
```
inline void AES_cfb_decrypt(AES_SRC_TYPE void *_in,
                            AES_DST_TYPE void *_out,
                            uint len, AES_KEY *akey, void *_iv)
{
        AES_SRC_TYPE uchar *in = _in;
        AES_DST_TYPE uchar *out = _out;
        uchar *iv = _iv;
        int n = 0;

        while (len--) {
                uchar c;

                if (!n)
                        AES_encrypt(iv, iv, akey);

                c = *in++;
                *out++ = c ^ iv[n];
                iv[n] = c;

                n = (n + 1) & 0x0f;
        }
}
```

**Comment by kholia** [comment]
@magnumripper I was able to implement a pure OpenCL kernel. Please review when you can. Thanks!

I had to resort to weird tricks (define new variables instead of reusing existing ones) to get the OpenCL kernel to work. There is not much logic behind these hacks it seems.

File: src/opencl_aes.h [hunk 1]
```diff
@@ -301,4 +301,48 @@ inline void AES_256_XTS_first_sector(AES_SRC_TYPE uint *in,
...
 }
 
+#define N_WORDS (AES_BLOCK_SIZE / sizeof(unsigned long))
+
+inline void
+AES_ige_decrypt(AES_CTS_SRC_TYPE void *_in, AES_CTS_DST_TYPE void *_out,
+                uint length, AES_KEY *akey, uchar *_iv)
+{
+	AES_CTS_SRC_TYPE uchar *in = _in;
+	AES_CTS_DST_TYPE uchar *out = _out;
```
Comment (magnumripper) [inline src/opencl_aes.h]: Please use `AES_SRC_TYPE` and `AES_DST_TYPE`. The more specific `AES_CTS_SRC_TYPE` was needed for some format that had different requirements at different stages.

**Comment by magnumripper** [comment]
> I had to resort to weird tricks (define new variables instead of reusing existing ones) to get the OpenCL kernel to work. There is not much logic behind these hacks it seems.

That's common, and very stupid (poor runtime shite). But the code is 100% fine and any decent optimizer will do the Right Thing anyway. So this is totally fine!

**Comment by kholia** [comment]
> Now I'm sure we could optimize this a tad, but that goes for several of my formats too.

I noticed byte-wise XOR'ing of `AES_BLOCK_SIZE` amount of data in some of the OpenCL AES code. This can be changed to XOR 32-bit chunks instead, but setting of the `union` based buffers initially might be costly.

This would be a tiny optimization as bulk of the time is consumed by the KDF in most cases (?).

**Comment by kholia** [comment]
> I knew you could do this! You are afraid of OpenCL for no reason at all.

Haha, yeah :-)

TIL that it's easy and pleasurable to work with the current OpenCL framework (and boilerplate code too!) we have in place now.

I learnt some debugging tricks as well when writing this kernel. This should give me more confidence regarding future OpenCL tasks.

**Comment by magnumripper** [comment]
> I noticed byte-wise XOR'ing of `AES_BLOCK_SIZE` amount of data in some of the OpenCL AES code. This can be changed to XOR 32-bit chunks instead, but setting of the `union` based buffers initially might be costly.

Unions cost nothing, not sure I understood what you meant there. Anyway most GPUs require aligned data, so if we're xor'ing 32-bits at a time, the input and output buffers would need to be aligned to 32 bits. Actually I'm hoping for the JIT build optimizers to end up with 32-bit code where applicable. I suspect some of them can do that.

**Comment by kholia** [comment]
> Unions cost nothing, not sure I understood what you meant there.

I saw some existing OpenCL code, which XOR's a chunk of data in a byte-wise fashion with a loop.

The same XOR can be done in 32-bit chunks using the union technique. However, setting up the union variable (buffer) will require a memcpy, which has its own cost. So, overall, I am not sure how much of a benefit it would be to move to explicit 32-bit XOR stuff using the union method.

**Pull request merged**